pkg/kritis/util/allowlist.go

```diff
@@ -39,20 +41,104 @@ func RemoveGloballyAllowedImages(images []string) []string {
 	return notAllowlisted
 }
 
-func imageInAllowlist(image string) (bool, error) {
-	for _, w := range constants.GlobalImageAllowlist {
-		allowlistRef, err := name.ParseReference(w, name.WeakValidation)
+// RemoveGloballyAllowedImages returns all images that aren't in gap allowlists
+func RemoveGapAllowedImages(images []string, allowlist []string) []string {
+	notAllowlisted := []string{}
+	for _, image := range images {
+		allowlisted, err := imageInGapAllowlist(image, allowlist)
+		if err != nil {
+			glog.Errorf("couldn't check if %s is in gap allowlist: %v", image, err)
+		}
+		if !allowlisted {
+			notAllowlisted = append(notAllowlisted, image)
+		}
+	}
+	return notAllowlisted
+}
+
+// Do an image match based on reference.
+// It checks whether the image and the pattern resolves to same URL,
+// e.g., gcr.io/kritis-project/preinstall.
+// Note that it does not check digest or tag in the pattern.
+// For example, a pattern of gcr.io/hello/world:latest will match any image
+// in the gcr.io/hello/world repository.
+func imageRefMatch(image string, pattern string) (bool, error) {
+	allowRef, err := name.ParseReference(pattern, name.WeakValidation)
+	if err != nil {
+		return false, err
+	}
+	imageRef, err := name.ParseReference(image, name.WeakValidation)
+	if err != nil {
+		return false, err
+	}
+	// Make sure c
+	if reflect.DeepEqual(allowRef.Context(), imageRef.Context()) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// Do an image match based on name pattern.
+// This method uses name pattern matching,
+// where a pattern is a path to a single image by
+// exact match, or to any images matching a pattern using the wildcard symbol
+// (`*`). The wildcards may only be present in the end, and not anywhere
+// else in the pattern, e.g., `gcr.io/n*x` is not allowed,
+// but `gcr.io/nginx*` is allowed. Also wilcards cannot be used to match `/`,
+// e.g., `gcr.io/nginx*` matches `gcr.io/nginx@latest`,
+// but it does not match `gcr.io/nginx/image`.
+// See more at https://cloud.google.com/binary-authorization/docs/policy-yaml-reference#admissionwhitelistpatterns
+func imageNamePatternMatch(image string, pattern string) (bool, error) {
+	if len(pattern) == 0 {
+		return false, errors.New("empty pattern")
+	}
+	if pattern[len(pattern)-1] == '*' {
+		pattern = pattern[:len(pattern)-1]
+		if strings.HasPrefix(image, pattern) &&
+			strings.LastIndex(image, "/") < len(pattern) {
+			return true, nil
+		}
+	} else {
+		if image == pattern {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func imageInAllowlistByReference(image string, allowList []string) (bool, error) {
```

Comment:
what does `Reference` term refer to here?

Reply:
See imageRefMatch(). I'll add some high level explanation.